lld/ELF/Arch/NanoMips.cpp

```diff
@@ -0,0 +1,562 @@
+template <endianness E> void writeValueHi20(uint8_t *loc, uint64_t val) {
+  uint32_t instr = readShuffle32<E>(loc);
+  uint32_t data = (val & ~1) | ((val >> 31) & 1);
+  data = (data & ~0xffc) | ((val >> 19) & 0xffc);
+  uint32_t masked = (instr & ~0x1ffffd) | (data & 0x1ffffd);
+  writeShuffle32<E>(loc, masked);
+}
+
+// used for: R_NANOMIPS_PC4_S1 and R_NANOMIPS_GPREL7_S2
+void writeValue16(uint8_t *loc, uint64_t val, uint8_t bitsSize, uint8_t shift) {
+  uint32_t instr = read16(loc);
+  uint32_t mask = (0xffff >> (16 - bitsSize)) << shift;
+  uint32_t data = (instr & ~mask) | (val & mask);
+  write16(loc, data);
+}
+
+// used for: R_NANOMIPS_PC10_S1 and R_NANOMIPS_PC7_S1
+void writePcRel16(uint8_t *loc, uint64_t val, uint8_t bitsSize) {
+  uint16_t instr = read16(loc);
+  val = (val & ~1) | ((val >> bitsSize) & 1);
+  uint16_t mask = (0xffff >> (16 - bitsSize));
+  uint16_t data = (instr & ~mask) | (val & mask);
+  write16(loc, data);
+}
+
+// used for: R_NANOMIPS_PC25_S1, R_NANOMIPS_PC21_S1, R_NANOMIPS_PC14_S1 and
+// R_NANOMIPS_PC11_S1
+template <endianness E>
+void writePcRel32(uint8_t *loc, uint64_t val, uint8_t bitsSize) {
+  uint32_t instr = readShuffle32<E>(loc);
+  val = (val & ~1) | ((val >> bitsSize) & 1);
+  uint32_t mask = (0xffffffff >> (32 - bitsSize));
+  uint32_t data = (instr & ~mask) | (val & mask);
+  writeShuffle32<E>(loc, data);
+}
+
+// used for: R_NANOMIPS_LO12, R_NANOMIPS_GPREL19_S2, R_NANOMIPS_GPREL18,
+// R_NANOMIPS_GPREL17_S1 and R_NANOMIPS_GPREL_LO12
+template <endianness E>
+void writeValue32Shifted(uint8_t *loc, uint64_t val, uint8_t bitsSize,
+                         uint8_t shift) {
+  uint32_t instr = readShuffle32<E>(loc);
+  uint32_t mask = (0xffffffff >> (32 - bitsSize)) << shift;
+  uint32_t data = (instr & ~mask) | (val & mask);
+  writeShuffle32<E>(loc, data);
+}
+// used for checking values of: pcrel relocations and gprel ones also
+// R_NANOMIPS_LO12
+void checkVal(uint8_t *loc, int64_t v, int n, const Relocation &rel,
+              const Symbol *sym, uint32_t shift, bool signedVal) {
+
+  if ((v & ((1 << shift) - 1)) != 0) {
+    error(getErrorLocation(loc) + "\tvalue: \t" + llvm::utohexstr(v) +
+          "\tlast N bits have to be zero in all S{N} relocations");
+  }
+
+  if (sym && sym->isUndefWeak())
+    return;
+  if (signedVal)
+    checkInt(loc, v, n, rel);
+  else
+    checkUInt(loc, v, n, rel);
+}
+
+bool isPlaceholderReloc(const RelType relType) {
+  switch (relType) {
+  case R_NANOMIPS_NONE:
+  case R_NANOMIPS_ALIGN:
+  case R_NANOMIPS_FILL:
+  case R_NANOMIPS_MAX:
+  case R_NANOMIPS_INSN32:
+  case R_NANOMIPS_FIXED:
+  case R_NANOMIPS_NORELAX:
+  case R_NANOMIPS_RELAX:
+  case R_NANOMIPS_SAVERESTORE:
+  case R_NANOMIPS_INSN16:
+  case R_NANOMIPS_JALR32:
+  case R_NANOMIPS_JALR16:
+  case R_NANOMIPS_JUMPTABLE_LOAD:
+    return true;
+  default:
+    return false;
+  }
+}
+
+// TODO: Support for other endianess, and bit size, now it is only Little Endian
+// 32 bit
```

Comment:
do we have an `assert` for this?

Reply:
There is no assert in NanoMips.cpp itself. However, in getTarget function of lld/ELF/Target.cpp if asked for any other target than 32-bit little endian llvm_unreachable is invoked.